core/src/main/java/com/linecorp/armeria/common/stream/StreamMessageInputStream.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.function.Function;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.internal.common.stream.StreamMessageUtil;
+
+final class StreamMessageInputStream<T> extends InputStream {
+
+    public static <T> StreamMessageInputStream<T> of(
+            StreamMessage<T> source, Function<? super T, ? extends HttpData> httpDataConverter) {
+        return new StreamMessageInputStream<>(source, httpDataConverter);
+    }
+
+    private final StreamMessageInputStreamSubscriber<T> subscriber;
+    @Nullable
+    private InputStream inputStream;
+
+    private StreamMessageInputStream(StreamMessage<T> source,
+                                     Function<? super T, ? extends HttpData> httpDataConverter) {
+        requireNonNull(source, "source");
+        requireNonNull(httpDataConverter, "httpDataConverter");
+        subscriber = new StreamMessageInputStreamSubscriber<>(httpDataConverter);
+        source.subscribe(subscriber);
+    }
+
+    @Override
+    public int read() throws IOException {
+        if (inputStream == null || inputStream.available() == 0) {
+            requestNextStream();
+        }
+        return inputStream.read();
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        if (inputStream == null || inputStream.available() == 0) {
+            requestNextStream();
+        }
+        return inputStream.read(b, off, len);
+    }
+
+    private void requestNextStream() {
+        subscriber.request();
+        try {
+            subscriber.latch.await();
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+        inputStream = subscriber.inputStream;
+    }
```

Comment:
![image](https://user-images.githubusercontent.com/34854527/155916122-864199d9-a58b-4b25-ae71-193395affc6a.png)
On `read()`, I used `CountDownLatch#await()` to **await** upstream(StreamMessage)'s `onNext()` is completed.

Cause I'm not familiar with this code(reactive streams..), maybe I'm in wrong direction.
Feel free to give me any suggestion! 🙇